sdks/python/apache_beam/io/range_trackers.py

```diff
@@ -100,15 +105,51 @@ def _validate_record_start(self, record_start, split_point):
           record_start)
 
   def try_claim(self, record_start):
+    assert not self.done()
     with self._lock:
+      if self.done():
+        raise ValueError('Trying to claim record at %r after already closing '
+                         'RangeTracker by invoking \'set_done()\'',
+                         record_start)
+
       self._validate_record_start(record_start, True)
       if record_start >= self.stop_position():
         return False
       self._offset_of_last_split_point = record_start
       self._last_record_start = record_start
+      self._split_points_seen += 1
+      if self._split_points_remaining_callback:
+        if (self._split_points_remaining ==
+            iobase.RangeTracker.SPLIT_POINTS_UNKNOWN):
+          if not self._first_claim_done:
+            # Trying to set the initial value for remaining number of split
+            # points.
+            # We only try to do this at first claim and when stop position gets
+            # updated so that we do not unnecessarily keep calling the callback
+            # for every call to try_claim().
+            self._split_points_remaining = (
+                self._split_points_remaining_callback(self.stop_position()))
+            if self._split_points_remaining <= 0:
+              raise ValueError(
```

Comment:
Note, BTW, that "remaining" is an estimate. Throwing an error if we were off by one would be bad.